src/main/java/org/jenkinsci/plugins/workflow/graph/StandardGraphLookupView.java

```diff
@@ -0,0 +1,237 @@
+                    String enclosingId = nearestEnclosingBlock.get(lookupId);
+                    if (enclosingId != null) {
+                        nearestEnclosingBlock.put(newHead.getId(), enclosingId);
+                    }
+                }
+            }
+        }
+    }
+
+    /** Create a lookup view for an execution */
+    public StandardGraphLookupView(FlowExecution execution) {
+        execution.addListener(this);
+    }
+
+    @Override
+    public boolean isActive(@Nonnull FlowNode node) {
+        if (node instanceof FlowEndNode) { // cf. JENKINS-26139
+            return node.getExecution().isComplete();
+        } else if (node instanceof BlockStartNode){  // BlockStartNode
+            return this.getEndNode((BlockStartNode)node) == null;
+        } else {
+            return node.getExecution().isCurrentHead(node);
+        }
+    }
+
+    // Do a brute-force scan for the block end matching the start, caching info along the way for future use
+    BlockEndNode bruteForceScanForEnd(@Nonnull BlockStartNode start) {
+        DepthFirstScanner scan = new DepthFirstScanner();
+        scan.setup(start.getExecution().getCurrentHeads());
+        for (FlowNode f : scan) {
+            if (f instanceof BlockEndNode) {
+                BlockEndNode end = (BlockEndNode)f;
+                BlockStartNode maybeStart = end.getStartNode();
+                // Cache start in case we need to scan again in the future
+                blockStartToEnd.put(maybeStart.getId(), end.getId());
+                if (start.equals(maybeStart)) {
+                    return end;
+                }
+            } else if (f instanceof BlockStartNode) {
+                BlockStartNode maybeThis = (BlockStartNode) f;
+
+                // We're walking from the end to the start and see the start without finding the end first, block is incomplete
+                String previousEnd = blockStartToEnd.get(maybeThis.getId());
+                if (previousEnd == null) {
+                    blockStartToEnd.put(maybeThis.getId(), INCOMPLETE);
+                }
+                if (start.equals(maybeThis)) {  // Early exit, the end can't be encountered before the start
+                    return null;
+                }
+            }
+        }
+        return null;
+    }
+
+
+
+
+    /** Do a brute-force scan for the enclosing blocks **/
+    BlockStartNode bruteForceScanForEnclosingBlock(@Nonnull final FlowNode node) {
+        FlowNode current = node;
+
+        while (!(current instanceof FlowStartNode)) {  // Hunt back for enclosing blocks, a potentially expensive operation
+            if (current instanceof BlockEndNode) {
+                // Hop over the block to the start
+                BlockStartNode start = ((BlockEndNode) current).getStartNode();
+                blockStartToEnd.put(start.getId(), current.getId());
+                current = start;
+                continue;  // Allows us to use this to handle special cases
```

Comment:
(I am not even trying to review this kind of code. Hope your test coverage is good.)

Reply:
@jglick Unfortunately that's not terribly helpful since this is the meat of the PR and you're the only other person familiar enough with this to really review it.   So, could I please ask you to give this a review -- seems only fair after I reviewed #45